extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -0,0 +1,606 @@
+    <div class="i-amphtml-story-expanded-view-overflow
+        i-amphtml-story-system-reset">
+      <span class="i-amphtml-expanded-view-close-button" role="button">
+        &times;
+      </span>
+    </div>`;
+
+/**
+ * Minimum vertical space needed to position tooltip.
+ * @const {number}
+ */
+const MIN_VERTICAL_SPACE = 56;
+
+/**
+ * Padding between tooltip and edges of screen.
+ * @const {number}
+ */
+const EDGE_PADDING = 8;
+
+/**
+ * Blank icon when no data-tooltip-icon src is specified.
+ * @const {string}
+ */
+const DEFAULT_ICON_SRC =
+  'data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=';
+
+/**
+ * @struct @typedef {{
+ *   tooltip: !Element,
+ *   buttonLeft: !Element,
+ *   buttonRight: !Element,
+ *   arrow: !Element,
+ * }}
+ */
+let tooltipElementsDef;
+
+const TAG = 'amp-story-embedded-component';
+
+const ComponentState = {
+  HIDDEN: 0, // Component is present in page, but hasn't been interacted with.
+  FOCUSED: 1, // Component has been clicked, a tooltip should be shown.
+  EXPANDED: 2, // Component is in expanded mode.
+};
+
+/**
+ * Embedded components found in amp-story.
+ */
+export class AmpStoryEmbeddedComponent {
+  /**
+   * @param {!Window} win
+   * @param {!Element} storyEl
+   */
+  constructor(win, storyEl) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.storyEl_ = storyEl;
+
+    /** @private {?Element} */
+    this.shadowRoot_ = null;
+
+    /** @private {?Element} */
+    this.focusedStateOverlay_ = null;
+
+    /** @private {?Element} */
+    this.tooltip_ = null;
+
+    /** @private {?Element} */
+    this.tooltipArrow_ = null;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = getStoreService(this.win_);
+
+    /** @private @const {!../../../src/service/resources-impl.Resources} */
+    this.resources_ = Services.resourcesForDoc(getAmpdoc(this.win_.document));
+
+    /** @private {?Element} */
+    this.expandedViewOverlay_ = null;
+
+    /**
+     * Target producing the tooltip. Used to avoid building the same
+     * element twice.
+     * @private {?Element} */
+    this.previousTarget_ = null;
+
+    /** @private */
+    this.expandComponentHandler_ = this.onExpandComponent_.bind(this);
+
+    this.storeService_.subscribe(StateProperty.EMBEDDED_COMPONENT, target => {
+      this.onEmbeddedComponentUpdate_(target);
+    });
+
+    this.state_ = ComponentState.HIDDEN;
+  }
+
+  /**
+   * Reacts to embedded component updates.
+   * @param {?Element} target
+   */
+  onEmbeddedComponentUpdate_(target) {
+    switch (this.state_) {
+      case ComponentState.FOCUSED:
+      case ComponentState.HIDDEN:
+        if (target) {
+          // Target is hidden and going into focused mode.
+          this.state_ = ComponentState.FOCUSED;
+
+        } else {
+          // Target is focused and click was outside the tooltip.
+          this.state_ = ComponentState.HIDDEN;
+        }
```

Comment:
Nit: when possible, trying to limit the number of indentation levels is always a good idea and helps readability a lot.
Here, we could remove the need for nesting a `if` in a `switch` by using the ternary operator:

`this.state_ = target ? ComponentState.FOCUSED : ComponentState.HIDDEN;`

Reply:
Done